### PR TITLE
WIP: Add timit recipe

### PR DESCRIPTION
**Pull request by luomingshuang**
Add timit recipe for icefall. This script uses phone as modeling units and it aims to compute the PER. Our target output is a list of phones. The split for {dev, test} is following kaldi ({[kaldi-timit-dev](https://github.com/kaldi-asr/kaldi/blob/master/egs/timit/s5/conf/dev_spk.list), [kaldi-timit-test](https://github.com/kaldi-asr/kaldi/blob/master/egs/timit/s5/conf/test_spk.list)}. At present, this script contains tdnn_lstm_ctc for use. And I will add other models and methods (such as comformer and crdnn, mmi) for it.

In fact, I have done some experiments for timit based on snowfall. https://github.com/k2-fsa/snowfall/pull/247

The current result is not the best. I will continue to improve it. 
[log-train-2021-10-28-15-24-21.txt](https://github.com/k2-fsa/icefall/files/7432206/log-train-2021-10-28-15-24-21.txt)
https://tensorboard.dev/experiment/twUbZTxoTAK32bPCJsYF7Q/#scalars

TODOs:
-  Add and check the scripts
-  Add documents for timit
-  Improve the performance

### CODE REVIEW

File: egs/timit/ASR/prepare.sh [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+# directories and files. If not, they will be downloaded
+# by this script automatically.
+#
+#  - $dl_dir/timit
+#      You can find data, train_data.csv, test_data.csv, etc, inside it.
+#      You can download them from https://data.deepai.org/timit.zip
+#
+#  - $dl_dir/lm
+#      This directory contains the language model(LM) downloaded from
+#      https://huggingface.co/luomingshuang/timit_lm, and the LM is based
```
Comment (csukuangfj) [inline egs/timit/ASR/prepare.sh:22]: Could you please describe how `lm_tgmed.arpa` is obtained?
Is it possible to train it inside icefall?

File: egs/timit/ASR/prepare.sh [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+   --lang-dir $lang_dir
+
+  if [ ! -f $lang_dir/L_disambig.pt ]; then
+    ./local/prepare_lang.py --lang-dir $lang_dir
+  fi
+fi
+
+if [ $stage -le 6 ] && [ $stop_stage -ge 6 ]; then
+  log "Stage 6: Prepare G"
+  # We assume you have install kaldilm, if not, please install
```
Comment (csukuangfj) [inline egs/timit/ASR/prepare.sh:125]: typo: install -> installed

File: egs/timit/ASR/prepare.sh [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+      $dl_dir/lm/lm_tgmed.arpa > data/lm/G_3_gram.fst.txt
+  fi
+
+  if [ ! -f data/lm/G_4_gram.fst.txt ]; then
+    # It is used for LM rescoring
+    python3 -m kaldilm \
+      --read-symbol-table="data/lang_phone/words.txt" \
+      --disambig-symbol='#0' \
+      --max-order=4 \
+      $dl_dir/lm/lm_tgmed.arpa > data/lm/G_4_gram.fst.txt
```
Comment (csukuangfj) [inline egs/timit/ASR/prepare.sh:144]: `tgmed` means this arpa is a tri-gram, of medium size, I think.
Please use a 4-gram arpa to generate `G_4_gram.fst.txt`, if you need it for decoding/rescoring.

File: egs/timit/ASR/shared/parse_options.sh [hunk 4]
```diff
@@ -0,0 +1,97 @@
+#!/usr/bin/env bash
```
Comment (csukuangfj) [inline egs/timit/ASR/shared/parse_options.sh:1]: This file is shared across various recipes.

Could you make it a symlink, like what we are doing in the librispeech recipe?

File: egs/timit/ASR/splits_dir/dev_samples.txt [hunk 5]
```diff
@@ -0,0 +1,400 @@
+FADG0_SI1279 TEST/DR4/FADG0/SI1279.WAV
```
Comment (csukuangfj) [inline egs/timit/ASR/splits_dir/dev_samples.txt:1]: Can this file be generated by some scripts? If so, we don't need to check it in.

**Review by csukuangfj — CHANGES_REQUESTED** [review]

File: egs/timit/ASR/local/prepare_lexicon.py [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+
+def prepare_lexicon(manifests_dir: str, lang_dir: str):
+    """
+    Args:
+      manifests_dir:
+        The manifests directory, e.g., data/manifests.
+      lang_dir:
+        The language directory, e.g., data/lang_phone.
+
+    Return:
```
[the remark this answers is not in the document]
Comment (csukuangfj) [reply egs/timit/ASR/local/prepare_lexicon.py:58]: By the way, it is `Returns`, not `Return`.

We are using google style python doc strings. Please have a look at
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html
when you have time.

File: egs/timit/ASR/prepare.sh (hunk 1 above)
Comment (luomingshuang) [reply egs/timit/ASR/prepare.sh:22]: Em....the lm_tgmed.arpa is obtained by this [train_lms.sh](https://github1s.com/luomingshuang/snowfall/blob/timit-recipe/egs/timit/asr/simple_v1/local2/train_lms.sh) which is followed by kaldi. About training lm inside icefall, I think it is a good idea. I have ever considered this problem that if we can train lm with python. There are some methods for it by using [KenLM](https://github.com/kpu/kenlm). Maybe I can have look.

File: egs/timit/ASR/RESULTS.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+2021-10-28 13:20:42,780 INFO [decode.py:360] Wrote detailed error stats to tdnn_lstm_ctc/exp/errs-TEST-lm_scale_1.8.txt
+2021-10-28 13:20:42,808 INFO [decode.py:351] The transcripts are stored in tdnn_lstm_ctc/exp/recogs-TEST-lm_scale_1.9.txt
+2021-10-28 13:20:42,828 INFO [utils.py:469] [TEST-lm_scale_1.9] %WER 33.35% [2406 / 7215, 40 ins, 1460 del, 906 sub ]
+2021-10-28 13:20:42,865 INFO [decode.py:360] Wrote detailed error stats to tdnn_lstm_ctc/exp/errs-TEST-lm_scale_1.9.txt
+2021-10-28 13:20:42,899 INFO [decode.py:351] The transcripts are stored in tdnn_lstm_ctc/exp/recogs-TEST-lm_scale_2.0.txt
+2021-10-28 13:20:42,919 INFO [utils.py:469] [TEST-lm_scale_2.0] %WER 33.97% [2451 / 7215, 39 ins, 1510 del, 902 sub ]
+2021-10-28 13:20:42,952 INFO [decode.py:360] Wrote detailed error stats to tdnn_lstm_ctc/exp/errs-TEST-lm_scale_2.0.txt
+2021-10-28 13:20:42,986 INFO [decode.py:374] 
+For TEST, PER of different settings are:
+lm_scale_0.1	20.82	best for TEST
```
[the remark this answers is not in the document]
Comment (luomingshuang) [reply egs/timit/ASR/RESULTS.md:171]: Doing.

File: egs/timit/ASR/local/compute_fbank_timit.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+        for partition, m in manifests.items():
+            if (output_dir / f"cuts_{partition}.json.gz").is_file():
+                logging.info(f"{partition} already exists - skipping.")
+                continue
+            logging.info(f"Processing {partition}")
+            cut_set = CutSet.from_manifests(
+                recordings=m["recordings"],
+                supervisions=m["supervisions"],
+            )
+            if "train" in partition:
```
[the remark this answers is not in the document]
Comment (luomingshuang) [reply egs/timit/ASR/local/compute_fbank_timit.py:73]: Oh....will do it....

File: egs/timit/ASR/local/prepare_lexicon.py [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+            text = load_dict["text"]
+            phones_list = list(filter(None, text.split(" ")))
+
+            for phone in phones_list:
+                if phone not in phones:
+                    phones.append(phone)
+
+    with open(lexicon, "w") as f:
+        for phone in sorted(phones):
+            f.write(str(phone) + "  " + str(phone))
```
[the remark this answers is not in the document]
Comment (luomingshuang) [reply egs/timit/ASR/local/prepare_lexicon.py:79]: OK.

File: egs/timit/ASR/prepare.sh (hunk 3 above)
Comment (luomingshuang) [reply egs/timit/ASR/prepare.sh:144]: Will do it.

File: egs/timit/ASR/prepare.sh (hunk 1 above)
Comment (csukuangfj) [reply egs/timit/ASR/prepare.sh:22]: ok, `train_lms.sh` uses https://github.com/danpovey/kaldi_lm.git

I will wrap it to Python with pybind11 when I have time.

File: egs/timit/ASR/splits_dir/dev_samples.txt (hunk 5 above)
Comment (luomingshuang) [reply egs/timit/ASR/splits_dir/dev_samples.txt:1]: Em....about {train, dev, test} spliting files, I don't find there are some scripts to generate them. In kaldi, there are placed in [a list file](https://github.com/kaldi-asr/kaldi/blob/master/egs/timit/s5/conf/dev_spk.list). In speechbrain, there are placed in [timit_prepare.py](https://github.com/speechbrain/speechbrain/blob/develop/recipes/TIMIT/timit_prepare.py) which listing the speakers with a list. A option for us is to learn speechbrain. We can use a list to contain speakers' names in data preparing process. I will add it to Lhotse.

**Pull request closed without merging**